crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -92,12 +97,17 @@ impl<'db> SemanticIndexBuilder<'db> {
             .expect("Always to have a root scope")
     }
 
-    fn push_scope(&mut self, node: NodeWithScopeRef) {
+    #[must_use]
+    fn push_scope(&mut self, node: NodeWithScopeRef) -> TryBlockContexts {
```

Comment:
I looked into `loop_break_states`, and I think it's actually fine; since every nested loop resets it, and (unlike try/except) there's no interaction with the parent loop state, there's no potential for scope-confusion bugs.

Reply:
> I don't think callers of `push_scope` and `pop_scope` should be required to keep track of try block contexts data returned from `push_scope` and pass it back into `pop_scope`. Instead we can keep a per-scope stack of TryBlockContexts on the builder itself; `push_scope` can push onto it and `pop_scope` can pop off of it, and it can all be handled transparently to the caller.

I'm okay with doing that, but how would that be different (in terms of memory usage) to keeping the `TryBlockContexts` state as a field on the `Scope` instances? I switched to this approach because of the concerns you stated in https://github.com/astral-sh/ruff/pull/13338#discussion_r1758151643

Reply:
It's different because Scope is permanent; it's stored in the SemanticIndex and used in type inference. We should only put data in it that is needed by type inference. 

Storing state on the SemanticIndexBuilder is ephemeral, only kept while we are building the semantic index for a given module, and then dropped when that's done.

Keeping the state on the builder as I've suggested here is in fact equivalent in memory usage to what we're doing right now in this PR; either way the same number of TryBlockContexts are alive at any given time. The only difference is whether their ownership is held on the stack (the current approach in this PR) or as part of the builder.

Reply:
Ahh, thanks, that makes a lot of sense!